lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -0,0 +1,567 @@
+                      SMTLIBSolver *SMTSolver,
+                      unsigned Timeout,
+                      bool &HasNextInputValue) {
+
+  HasNextInputValue = true;
+  Inst *Ante = IC.getConst(APInt(1, true));
+  for (auto PC : PCs ) {
+    Inst* Eq = IC.getInst(Inst::Eq, 1, {PC.LHS, PC.RHS});
+    Ante = IC.getInst(Inst::And, 1, {Ante, Eq});
+  }
+
+  // If a variable is neither found in PCs or TriedVar, return APInt(0)
+  if (TriedVars.find(Var) == TriedVars.end()) {
+    std::vector<Inst *> VarsInPCs;
+    findVars(Ante, VarsInPCs);
+    if (std::find(VarsInPCs.begin(), VarsInPCs.end(), Var) == VarsInPCs.end()) {
+      TriedVars[Var].push_back(APInt(Var->Width, 0));
+      return APInt(Var->Width, 0);
+    }
+  }
+
+  for (auto Value : TriedVars[Var]) {
+    Inst* Ne = IC.getInst(Inst::Ne, 1, {Var, IC.getConst(Value)});
+    Ante = IC.getInst(Inst::And, 1, {Ante, Ne});
+  }
+
+  InstMapping Mapping(Ante, IC.getConst(APInt(1, true)));
+
+  std::vector<Inst *> ModelInsts;
+  std::vector<llvm::APInt> ModelVals;
+  std::string Query = BuildQuery(IC, {}, {}, Mapping, &ModelInsts, /*Negate=*/ true);
+
+  bool PCQueryIsSat;
+  std::error_code EC;
+  EC = SMTSolver->isSatisfiable(Query, PCQueryIsSat, ModelInsts.size(), &ModelVals, Timeout);
+
+  if (EC || !PCQueryIsSat) {
+    HasNextInputValue = false;
+    return APInt(Var->Width, 0);
+  }
+  if (DebugLevel > 2)
+    llvm::errs() << "Input guess SAT\n";
+  for (unsigned I = 0 ; I != ModelInsts.size(); I ++) {
+    if (ModelInsts[I] == Var) {
+      TriedVars[Var].push_back(ModelVals[I]);
+      if (DebugLevel > 2)
+        llvm::errs() << "Guess input value = " << ModelVals[I] << "\n";
+      return ModelVals[I];
+    }
+  }
+
+  llvm::report_fatal_error("Model does not contain the guess input variable");
+}
+
+std::error_code
+ExhaustiveSynthesis::synthesize(SMTLIBSolver *SMTSolver,
+                                const BlockPCs &BPCs,
+                                const std::vector<InstMapping> &PCs,
+                                Inst *LHS, Inst *&RHS,
+                                InstContext &IC, unsigned Timeout) {
+  std::vector<Inst *> Vars;
+  findVars(LHS, Vars);
+  std::vector<Inst *> Inputs;
+  // TODO tune the number of candidate inputs
+  findCands(LHS, Inputs, /*WidthMustMatch=*/false, /*FilterVars=*/false, 15);
+
+  llvm::errs() << "LHS has " << Vars.size() << " vars\n";
+
+  std::vector<Inst *> Guesses;
+  int LHSCost = souper::cost(LHS);
+  getGuesses(Guesses, Vars, Inputs, LHS->Width, LHSCost, IC);
+
+  std::error_code EC;
+  if (Guesses.size() < 1)
+    return EC;
+
+  // one of the real advantages of this approach to synthesis vs
+  // CEGIS is that we can synthesize in precisely increasing cost
+  // order, and not try to somehow teach the solver how to do that
+  std::sort(Guesses.begin(), Guesses.end(),
+            [](Inst *a, Inst *b) -> bool {
+              return souper::cost(a) < souper::cost(b);
+            });
+
+  // Big Query
```

Comment:
please add a TODO that we need to check how many times the big query saves us time, instead of just wasting time